client/plugins.c

```diff
@@ -0,0 +1,471 @@
+/*
+ * Copyright (C) 2020 VMware, Inc. All Rights Reserved.
+ *
+ * Licensed under the GNU Lesser General Public License v2.1 (the "License");
+ * you may not use this file except in compliance with the License. The terms
+ * of the License are located in the COPYING file of this distribution.
+ */
+
+#include "includes.h"
+#include <dlfcn.h>
+
+struct plugin_config
+{
+    PTDNF_CMD_ARGS pArgs;
+    char *pszPath;
+    char *pszConfPath;
+};
+
+static
+uint32_t
+_TDNFLoadPlugins(
+    PTDNF_CMD_ARGS pArgs,
+    PTDNF_PLUGIN *ppPlugins
+    );
+
+/*
+ * plugins are c libraries which are dynamically loaded
+ * if noplugins is set, this function returns immediately
+ * without further processing
+ * files with extension .conf are enumerated from the config
+ * directory.
+ * If a plugin is disabled by command line override,
+ * the config is skipped.
+ * If a plugin's config file has the enabled flag set to 0,
+ * the config is skipped.
+ * For all the configs that are parsed without skipping,
+ * the corresponding shared library is dynamically loaded
+ * as lib<pluginname>.so from the plugin dir where <pluginname>
+ * is the file name (without extension) of the corresponding
+ * config file.
+*/
+uint32_t
+TDNFLoadPlugins(
+    PTDNF pTdnf
+    )
+{
+    uint32_t dwError = 0;
+    PTDNF_PLUGIN pPlugins = NULL;
+
+    if (!pTdnf || !pTdnf->pArgs)
+    {
+        dwError = ERROR_TDNF_INVALID_PARAMETER;
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    dwError = _TDNFLoadPlugins(pTdnf->pArgs, &pPlugins);
+    BAIL_ON_TDNF_ERROR(dwError);
+
+    pTdnf->pPlugins = pPlugins;
+
+cleanup:
+    return dwError;
+
+error:
+    if (dwError == ERROR_TDNF_PLUGINS_DISABLED ||
+        dwError == ERROR_TDNF_NO_PLUGIN_CONF_DIR)
+    {
+        dwError = 0;
+    }
+    goto cleanup;
+}
+
+static
+uint32_t
+_TDNFGetPluginSettings(
+    struct plugin_config *pConf
+    )
+{
+    uint32_t dwError = 0;
+    int nHasOpt = 0;
+
+    dwError = TDNFHasOpt(pConf->pArgs, TDNF_CONF_KEY_NO_PLUGINS, &nHasOpt);
+    BAIL_ON_TDNF_ERROR(dwError);
+
+    if (nHasOpt)
+    {
+        dwError = ERROR_TDNF_PLUGINS_DISABLED;
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    dwError = TDNFGetOptWithDefault(
+                  pConf->pArgs, TDNF_CONF_KEY_PLUGIN_CONF_PATH,
+                  TDNF_DEFAULT_PLUGIN_CONF_PATH, &pConf->pszConfPath);
+    BAIL_ON_TDNF_ERROR(dwError);
+
+    dwError = TDNFGetOptWithDefault(
+                  pConf->pArgs, TDNF_CONF_KEY_PLUGIN_PATH,
+                  TDNF_DEFAULT_PLUGIN_CONF_PATH, &pConf->pszPath);
```

Comment:
Should this be  TDNF_DEFAULT_PLUGIN_PATH instead of TDNF_DEFAULT_PLUGIN_CONF_PATH?

Reply:
Good catch. Fixed.